src/main/java/com/patchworkmc/event/EventHandlerRewriter.java

```diff
@@ -68,34 +101,79 @@ public MethodVisitor visitMethod(int access, String name, String descriptor, Str
+		MethodVisitor instanceRegistrar = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC, EventConstants.REGISTER_INSTANCE, EventConstants.getRegisterInstanceDesc(className), null, null);
+		instanceRegistrar.visitCode();
+
+		// Check the target instance isn't null.
+		// TODO: this potentially causes differing bytecode when recompiled, because
+		// TODO: javac will insert this check for every lambda and some decompilers (cough fernflower) don't properly detect the J9 method.
+
+		instanceRegistrar.visitVarInsn(Opcodes.ALOAD, 0); // Load the target instance for null-checking (1)
+		// In every java version this is run for each INVOKEDYNAMIC but we only do it once here.
+		// In java 8, this is a INVOKEVIRTUAL Object.getClass but we use the Java 9 way here instead because it makes more sense
+		// pops the instance and instantly returns it (1)
+		instanceRegistrar.visitMethodInsn(Opcodes.INVOKESTATIC, "java/util/Objects", "requireNonNull", "(Ljava/lang/Object;)Ljava/lang/Object;", false);
+		instanceRegistrar.visitInsn(Opcodes.POP); // pop the returned instance (0)
 
-		if (staticRegistrar != null) {
-			staticRegistrar.visitCode();
+		// Register the subscribers
 
-			for (SubscribeEvent subscriber: staticSubscribers) {
-				Handle handler = new Handle(Opcodes.H_INVOKEVIRTUAL, className, subscriber.getMethod(), subscriber.getMethodDescriptor(), false);
+		for (SubscribeEvent subscriber : instanceSubscribeEvents) {
+			instanceRegistrar.visitVarInsn(Opcodes.ALOAD, 1); // Load IEventBus on to the stack (1)
+			instanceRegistrar.visitVarInsn(Opcodes.ALOAD, 0); // Load the target instance (2)
 
-				staticRegistrar.visitVarInsn(Opcodes.ALOAD, 0);
-				staticRegistrar.visitInvokeDynamicInsn("accept", "()Ljava/util/function/Consumer;", metafactory, Type.getMethodType("(Ljava/lang/Object;)V"), handler, Type.getMethodType(subscriber.getMethodDescriptor()));
-				staticRegistrar.visitMethodInsn(Opcodes.INVOKEINTERFACE, EVENT_BUS, "addListener", "(Ljava/util/function/Consumer;)V", true);
+			int callingOpcode;
+
+			if (isInterface) {
+				callingOpcode = Opcodes.H_INVOKEINTERFACE;
+			} else if (isFinalClass || (subscriber.getAccess() & Opcodes.ACC_FINAL) != 0) {
+				callingOpcode = Opcodes.H_INVOKESPECIAL;
+			} else {
+				callingOpcode = Opcodes.H_INVOKEVIRTUAL;
 			}
 
-			staticRegistrar.visitInsn(Opcodes.RETURN);
-			staticRegistrar.visitMaxs(2, 1);
-			staticRegistrar.visitEnd();
+			// Swap the target instance with a Consumer instance (2)
+			LambdaUtil.visitConsumerInstanceMethodReference(instanceRegistrar, callingOpcode, className, subscriber.getMethod(), subscriber.getMethodDescriptor(), isInterface);
+
+			// Pop the eventbus instance and the lambda. (0)
+			instanceRegistrar.visitMethodInsn(Opcodes.INVOKEINTERFACE, EventConstants.EVENT_BUS, "addListener", "(Ljava/util/function/Consumer;)V", true);
```

Comment:
This doesn't use addGenericListener at all, neither does genStaticRegistrar. Can we at least get some TODOs?